src/main/java/org/traccar/geocoder/AmapGeocoder.java

```diff
@@ -0,0 +1,98 @@
+/*
+ * Copyright 2012 - 2017 Anton Tananaev (anton@traccar.org)
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.traccar.geocoder;
+
+import jakarta.json.JsonArray;
+import jakarta.json.JsonObject;
+import jakarta.json.JsonString;
+import jakarta.json.JsonValue;
+import jakarta.ws.rs.client.Client;
+import org.traccar.helper.CoordinateUtil;
+
+
+public class AmapGeocoder extends JsonGeocoder {
+
+    private static String formatUrl(String key) {
+        String url = "https://restapi.amap.com/v3/geocode/regeo?output=json&location=%2$f,%1$f&extensions=all";
+        if (key != null) {
+            url += "&key=" + key;
+        }
+        return url;
+    }
+
+    public AmapGeocoder(Client client, String key, int cacheSize, AddressFormat addressFormat) {
+        super(client, formatUrl(key), cacheSize, addressFormat);
+    }
+
+    public Address parseAddress(JsonObject json) {
+        JsonObject regeocode = json.getJsonObject("regeocode");
+        if (regeocode != null) {
+            Address address = new Address();
+
+            address.setFormattedAddress(getJsonNode(regeocode, "formatted_address"));
+
+            JsonObject addressComponent = regeocode.getJsonObject("addressComponent");
+            if (addressComponent != null) {
+                JsonObject building = addressComponent.getJsonObject("building");
+                String houseName = getJsonNode(building, "name");
+                if (houseName == null) {
+                    JsonArray aois = regeocode.getJsonArray("aois");
+                    if (!aois.isEmpty()) {
+                        JsonObject aoi = aois.getJsonObject(0);
+                        houseName = getJsonNode(aoi, "name");
+                    }
+                }
+                address.setHouse(houseName);
+
+                JsonObject street = addressComponent.getJsonObject("streetNumber");
+                address.setStreet(getJsonNode(street, "street"));
+
+                address.setSuburb(getJsonNode(addressComponent, "township"));
+                address.setSettlement(getJsonNode(addressComponent, "district"));
+                address.setDistrict(getJsonNode(addressComponent, "city"));
```

Comment:
This doesn't look right. Can you please provide documentation for the API.

Reply:
English is not my mother tongue. Therefore, my understanding of some English words may be deficient. I really do not quite understand the scope of the area expressed by the word "settlement". Usually in China, the order of area scope is: country, province, city, district, town, street and building. But I read the comment on the "AddressFormat" class, and "district" comes before "settlement". I think "district" expresses a larger scope than "settlement". So I think "settlement" for "district" and "district" for "city" may be more compatible.

Reply:
Do you have the API documentation?

Reply:
Yes, the API documentation can be accessed at https://lbs.amap.com/api/webservice/guide/api/georegeo/#t5 . However, this documentation is only available in Chinese.

Reply:
If I understand it correctly, city in the API is only large cities that are basically their own "counties". I think the mapping should be:

country -> country
province -> state
district -> district
township -> settlement
neighborhood -> suburb

Reply:
This understanding is not quite correct. In China, "city" is a very important administrative unit. It is even more important than "province" in the address.
For example, Guangdong is a province, Guangzhou is a city under Guangdong, and Tianhe is a district under Guangzhou.
In address writing, we would use "Tianhe, Guangzhou, Guangdong, China" for Tianhe District, or "Tianhe, Guangzhou, China "But we will not use "Tianhe, Guangdong, China" for Tianhe District.
However, in some special cases, "city" can also stand for "province". There are four direct-administered municipality in China, which are provincial administrative regions together with provinces. They are: Beijing, Tianjin, Shanghai and Chongqing.

A "neighbourhood" is a very small administrative unit under "township". In most cases, we don't use "neighbourhood" in addresses. Most Chinese people don't even know which "neighbourhood" they belong to.

Reply:
OK, but city only applies to the places where there is a big city an administrative area. So it's not always there.

The mapping should be based on the logical equivalents, not on the order.

Reply:
Maybe you don't live in China, so you don't understand the importance of "city". In most cases, addresses in China use "city". For historical reasons, there are very few direct-administered counties in China, less than one-twentieth of the number of cities. Therefore, the unit "city" is very important and indispensable.

Reply:
Then we can potentially do something like `district = if city == null ? district : city`

Reply:
In a few cases, it is possible to express it this way. But in most cases, it is necessary to keep both "city" and "district" in the address, because in more cases, they are upper and lower levels rather than the same level.
And "neighbourhood" is an unnecessary unit. Because addresses in China almost never use "neighbourhood".
You can refer to the address of the U.S. Embassy in China: https://china.usembassy-china.org.cn/contact/ . The address basically only contains building (number), street/road, township, district, city and province.
Or you can also refer to SAP, the software company's documentation for the Chinese address format: https://help.sap.com/docs/SAP_DATA_QUALITY_MANAGEMENT_SDK/7ac66320ca514acc89396a367db6dba8/ e540dc567ee51014a997d681b0e91070.html?version=4.2.9&locale=en-US

Translated with www.DeepL.com/Translator (free version)

Reply:
In a few cases, it is possible to express it this way. But in most cases, it is necessary to keep both "city" and "district" in the address, because in more cases, they are upper and lower levels rather than the same level.
And "neighbourhood" is an unnecessary unit. Because addresses in China almost never use "neighbourhood".
You can refer to the address of the U.S. Embassy in China: https://china.usembassy-china.org.cn/contact/ . The address basically only contains building (number), street/road, township, district, city and province.
Or you can also refer to SAP, the software company's documentation for the Chinese address format: https://help.sap.com/docs/SAP_DATA_QUALITY_MANAGEMENT_SDK/7ac66320ca514acc89396a367db6dba8/e540dc567ee51014a997d681b0e91070.html?version=4.2.9&locale=en-US

Reply:
<img width="555" alt="image" src="https://github.com/traccar/traccar/assets/1845381/91fe4d6a-dddb-4795-a87b-4722a0d515d0">
China's addressing system is very complex. This is the largest private courier company in China, the function of filling in the mail delivery address, with high reference.

Reply:
After checking some information, I found that the address expression form in the United States is really very different from the address expression form in China. For example, in the United States, "Settlement" can mean a village or a city, and "District" can mean a part of a city or several cities. There is no clear hierarchy between "settlement" and "district". In China, however, "District" is a smaller unit under "City".
So in the end, I think this is a more appropriate mapping.
country -> country
province -> state
district -> district
city -> settlement
township -> suburb